staging/src/k8s.io/legacy-cloud-providers/aws/aws.go

```diff
@@ -3860,21 +3860,33 @@ func (c *Cloud) EnsureLoadBalancer(ctx context.Context, clusterName string, apiS
 		}
 	} else {
 		klog.V(4).Infof("service %v does not need custom health checks", apiService.Name)
-		// We only configure a TCP health-check on the first port
-		var tcpHealthCheckPort int32
-		for _, listener := range listeners {
-			if listener.InstancePort == nil {
-				continue
-			}
-			tcpHealthCheckPort = int32(*listener.InstancePort)
-			break
-		}
 		annotationProtocol := strings.ToLower(annotations[ServiceAnnotationLoadBalancerBEProtocol])
 		var hcProtocol string
+		var tcpHealthCheckPort int32
 		if annotationProtocol == "https" || annotationProtocol == "ssl" {
 			hcProtocol = "SSL"
+			for _, listener := range listeners {
+				if listener.InstancePort == nil {
+					continue
+				}
+				protocol := listener.Protocol
+				if *protocol == "HTTPS" || *protocol == "SSL" {
+					tcpHealthCheckPort = int32(*listener.InstancePort)
+					break
+				}
+			}
+			if tcpHealthCheckPort == 0 {
```

Comment:
Will this break current behavior?  Should we fall back to picking the first port here too instead of erroring?

Reply:
I'm thinking its actually ok, because we're making the assumption that the backend protocol annotation dictates the health check protocol, and AFAIK the only time a user should choose a secure backend protocol is if all their instance ports are secure.  (According to the [k8s aws service loadbalancer docs](https://kubernetes.io/docs/concepts/services-networking/service/#ssl-support-on-aws), when using mixed ports the user should select the insecure protocol for backend protocol).  @justinsb @M00nF1sh can you confirm that its ok to fail when backend protocol annotation is secure and there are no ports with a *secure frontend protocol*?